packages/react-native-tracer-provider/src/EmbraceNativeSpan.ts

```diff
@@ -0,0 +1,224 @@
+  ) {
+    this.tracerName = tracerName;
+    this.tracerVersion = tracerVersion;
+    this.createdIndex = createdIndex;
+    this.spanContextSyncBehaviour = spanContextSyncBehaviour;
+  }
+
+  public nativeID(): string {
+    return `${this.tracerName}_${this.tracerVersion}_${this.createdIndex}`;
+  }
+
+  public creatingNativeSide(creating: Promise<SpanContext>) {
+    this.creating = creating;
+    this.creating.then((spanContext: SpanContext) => {
+      this.savedSpanContext = spanContext;
+    });
+  }
+
+  public isReadonly() {
+    return !this.recording;
+  }
+
+  /**
+   * The fact that native module calls are asynchronous and the @opentelemetry/api is synchronous doesn't matter for
+   * most methods as they are generally fire and forget. Getting spanContext is an exception since this could be asked
+   * for immediately after having started a span and if the value is not yet ready on the native side we will either:
+   *  - do a blocking synchronous call to retrieve it
+   *  - throw an error
+   *  - return a blank span context
+   *
+   * Behaviour can be figured on the EmbraceNativeTracerProvider by supplying spanContextSyncBehaviour
+   */
+  public spanContext(): SpanContext {
+    if (this.savedSpanContext) {
+      return this.savedSpanContext;
+    }
+
+    switch (this.spanContextSyncBehaviour) {
+      case 'block':
+        return TracerProviderModule.spanContext(this.nativeID());
+      case 'throw':
+        throw new Error('span context is not yet available on the native side');
+      case 'return_empty':
+        return {
+          traceId: '',
+          spanId: '',
+          traceFlags: 0,
+        };
+    }
+  }
+
+  /**
+   * spanContextAsync provides an async alternative to spanContext by wrapping the response in a promise
+   */
+  public spanContextAsync(): Promise<SpanContext> {
+    return this.creating || Promise.reject('failed to retrieve span context');
+  }
+
+  public setAttribute(key: string, value: AttributeValue): this {
+    return this.setAttributes({ [key]: value });
+  }
+
+  public setAttributes(attributes: Attributes): this {
+    if (this.isReadonly()) {
+      return this;
+    }
+    TracerProviderModule.setAttributes(this.nativeID(), normalizeAttributes(attributes));
+    return this;
+  }
+
+  public addEvent(
+    name: string,
+    attributesOrStartTime?: Attributes | TimeInput,
+    timeStamp?: TimeInput
+  ): this {
+    if (this.isReadonly()) {
+      return this;
+    }
+
+    if (isAttributes(attributesOrStartTime)) {
+      TracerProviderModule.addEvent(
+        this.nativeID(),
+        name,
+        normalizeAttributes(attributesOrStartTime),
+        normalizeTime(timeStamp)
+      );
+    } else {
+      TracerProviderModule.addEvent(
+        this.nativeID(),
+        name,
+        normalizeAttributes({}),
+        normalizeTime(attributesOrStartTime)
+      );
+    }
+
+    return this;
+  }
+
+  public addLink(link: Link): this {
+    return this.addLinks([link]);
```

Comment:
Could we save an allocation on boxing the link in an array by inverting `addLinks` and `addLink`? That is, by writing this such that `addLinks` calls `addLink` rather than this way around?

Reply:
Makes sense. Thanks for clarifying!